interp/interp_test.go

```diff
@@ -191,3 +190,341 @@ func TestPiecewiseConstantPredict(t *testing.T) {
 	leftYs := []float64{-0.5, 1.5, 1.5, 1.5, 1, 1}
 	testInterpolatorPredict(t, pc, testXs, leftYs, 0)
 }
+
+func TestPiecewiseCubic(t *testing.T) {
+	t.Parallel()
+	const (
+		h        = 1e-8
+		valueTol = 1e-13
+		derivTol = 1e-6
+	)
+	for i, test := range []struct {
+		xs []float64
+		f  func(float64) float64
+		df func(float64) float64
+	}{
+		{
+			xs: []float64{-1.001, 0.2, 2},
+			f:  func(x float64) float64 { return x * x },
+			df: func(x float64) float64 { return 2 * x },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 2},
+			f:  func(x float64) float64 { return x * x },
+			df: func(x float64) float64 { return 2 * x },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 10},
+			f:  func(x float64) float64 { return 1.5*x - 1 },
+			df: func(x float64) float64 { return 1.5 },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 10},
+			f:  func(x float64) float64 { return -1 },
+			df: func(x float64) float64 { return 0 },
+		},
+		{
+			xs: []float64{-1.1, 0.2, 0.99, 2.5, 2.99},
+			f:  math.Cos,
```

Comment:
If you use either only at most 3. degree polynomials or if you add a flag to indicate that a test case is such, you could check the fit (which should be very good) also at points that are not knots.

Reply:
Good point.